good-enough-practices-for-scientific-computing.tex

```diff
@@ -434,32 +449,52 @@ \subsection*{Acknowledgments}
 
 \item
   \practice{2b}{Decompose programs into functions} that are no more than
-  one page (about 60 lines) long, do not
-  use global variables (constants are OK), and take no more than half
-  a dozen parameters.  The key motivation here is to fit the program
+  one page (about 60 lines) long. Functions are sections within a program
+  that do one particular thing. The syntax varies depending on programming
+  language, but generally you name the function, list the parameters, or 
+  information needed for that section of code, and list what information will 
+  come out of the function. 
+  % check definition of functions
+  % WOULD BE GOOD TO HAVE EXAMPLE OF A FUNCTION
+  Functions should take no more than about 5 or 6 input parameters
+  and shouldn't include variables used in other parts of the program 
+  (global variables).
+  % I'm not sure if we should discuss global variables. I'm not sure it's
+  % an easy concept to explain. Or better description of global variables here?
+  The key motivation here is to fit the program
   into the most limited memory of all: ours. Human short-term memory
   is famously incapable of holding more than about seven items at
   once \cite{miller1956}. If we are to understand
   what our software is doing, we must break it into chunks that obey
-  this limit, then create programs by combining these chunks.
+  this limit, then create programs by combining these chunks. Putting 
+  code into functions also makes it easier to test and troubleshoot if
+  something is going wrong. 
 
 \item
   \practice{2c}{Be ruthless about eliminating duplication}. Write and
-  re-use functions instead of copying and pasting source code, and use
-  data structures like lists rather than creating lots of variables
-  called \texttt{score1}, \texttt{score2}, \texttt{score3}, etc.
-
-  The easiest code to debug and maintain is code you didn't
-  actually write, so \practice{2d}{always search for well-maintained
+  re-use functions instead of copying and pasting code. You can 
+  also use 
+  data structures like lists, e.g. create \texttt{score = (1, 2, 3)} rather than lots of variables
+  like \texttt{score1 = 1}, \texttt{score2 = 2}, \texttt{score3 = 3}, etc.
+
+  Also look for well-maintained libraries that already do what you're trying to do. All
+  programming languages have libraries that you can import and 
+  use in your code. This is code that people have already written and made available
+  for distribution that have a particular function. For instances there are libraries for statistics,
+  modeling, mapping and many more. Many languages catalog the libraries in a centralized
+  source, for instance R has CRAN, Python has ... and Java has ...
```

Comment:
I did first have Perl in instead of Java. :) We could still include it if we wanted more examples. I don't think it's as widely used, even in bioinformatics now though.